iohk/common/Arch.hs

```diff
@@ -0,0 +1,99 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE DeriveFoldable #-}
+{-# LANGUAGE DeriveTraversable #-}
+
+module Arch
+  ( Arch(..)
+  , ArchMap
+  , ApplicationVersionKey
+  , formatArch
+  , archMap
+  , mkArchMap
+  , archMapToList
+  , archMapFromList
+  , lookupArch
+  , idArchMap
+  , archMapEach
+  ) where
+
+import qualified Data.Aeson            as AE
+import           Data.Aeson
+import           Data.Text              (Text)
+import           GHC.Generics
+
+-- * Arch
+data Arch = Linux64 | Mac64 | Win64 deriving (Show, Read, Eq, Generic)
+
+instance FromJSON Arch
+instance ToJSON Arch
+
+type ApplicationVersionKey = Arch -> Text
+
+formatArch :: Arch -> Text
+formatArch Linux64 = "Linux"
+formatArch Mac64 = "macOS"
+formatArch Win64 = "Windows"
+
+-- | A map of values indexed by Arch.
+-- Maybe "type ArchMap a = Arch -> a" would be better but this works.
```

Comment:
Could this state *why* `Arch -> a` would be better, or maybe just omit the comment?

My reasoning is that the next person reading this code will wonder why the original
author left that comment.

Reply:
OK removed the confusing comment.

An assoc list or Map could have missing keys but ArchMap can't.

Really it doesn't matter too much about the implementation. This one might be more boilerplate than necessary but it's fairly straightforward.